### PR TITLE
Us36

### PR DESCRIPTION
**Pull request by LelandCurtis**
- items ready to ship
- create new factories
- fix bugs with updated test data.

### CODE REVIEW

File: app/models/customer.rb [hunk 1]
```diff
@@ -1,11 +1,16 @@
 class Customer < ApplicationRecord
   has_many :invoices
+  has_many :transactions, through: :invoices
```
Comment (jaw772) [inline app/models/customer.rb:3]: Brilliant!

**Review by jaw772 — APPROVED** [review]

File: app/models/customer.rb (hunk 1 above)
Comment (EMielke76) [reply app/models/customer.rb:3]: I honestly didn't know this was possible. I remember you asking early on if this was a thing we could do, super cool to see that it is :D Well done!

File: app/models/customer.rb [hunk 2]
```diff
@@ -1,11 +1,16 @@
...
   validates :first_name, presence: true
   validates :last_name, presence: true
 
   def successful_transactions_count
-    total = invoices.joins(:transactions).where(:transactions => {result: 0}).count
+    successful_transactions.count
+  end
+
+  def successful_transactions
+    transactions.where(result: 0)
```
Comment (EMielke76) [inline app/models/customer.rb:13]: Think smarter, not harder. Super cool to see you using the association you created right away in here!

File: spec/factories/transactions.rb [hunk 3]
```diff
@@ -4,5 +4,14 @@
     sequence(:credit_card_number) { |n| (1234567890123456 + n).to_s }
     credit_card_expiration_date { "2022-01-03" }
     result { 0 }
+
+    factory :transaction_with_customer do
+      transient do
+        customer {create(:customer)}
+      end
+      after(:create) do |transaction, evaluator|
+        transaction.invoice.update(customer: evaluator.customer)
```
Comment (EMielke76) [inline spec/factories/transactions.rb:13]: Thiiiis is neat! This is calling the created transaction's invoice, then updating that invoices customer information?

File: spec/features/merchants/dashboard_spec.rb [hunk 4]
```diff
@@ -21,75 +21,104 @@
...
+  describe 'top 5 customers section' do
+    it "has the names of the top 5 customers with largest number of completed transactions" do
+      merchant = create(:merchant)
+
+      customer_1 = create(:customer_with_transactions, merchant: merchant, transaction_result: 0, transaction_count: 6, first_name: 'Bob')
+      customer_2 = create(:customer_with_transactions, merchant: merchant, transaction_result: 0, transaction_count: 3, first_name: 'John')
+      customer_3 = create(:customer_with_transactions, merchant: merchant, transaction_result: 0, transaction_count: 8, first_name: 'Abe')
+      customer_4 = create(:customer_with_transactions, merchant: merchant, transaction_result: 0, transaction_count: 1, first_name: 'Zach')
+      customer_5 = create(:customer_with_transactions, merchant: merchant, transaction_result: 0, transaction_count: 4, first_name: 'Charlie')
+
```
Comment (EMielke76) [inline spec/features/merchants/dashboard_spec.rb:33]: The power of factory bot on full display! Thank you for taking the time over break to learn that, and for bringing that knowledge to us!! So badass to see it flex it's muscle in this way!

**Review by EMielke76 — COMMENTED** [review]
So cool to see how grand factory bot can go. Just trying to think of writing all those transactions by hand makes my brain hurt. 

Well done!

**Review by EMielke76 — APPROVED** [review]

**Pull request merged**